### ISSUE TITLE
Append operation on multi-relationships of subscribed entities does not trigger notifications

### ISSUE THREAD

**Issue by daniel-gonzalez-sanchez** [ScorpioBroker/ScorpioBroker#219]
Hi everyone,

I have discover another possible issue regarding NGSI-LD notifications from entities with one-to-many relationships or multi-relationships (it is connected to a previous closed issue - #205). I've checked that issuing an Append operation on the 
multi-relationship of a subscribed entity does not trigger notifications.

Let me show an example with the same use case I followed in the related previous issue.

**Environment**

- scorpiobroker/scorpio:scorpio-aaio_FIWARE_7.9.2

**Steps to reproduce**

1. Create subscription to a `House` entity.
```
curl --location --request POST 'localhost:9090/ngsi-ld/v1/subscriptions/' \
--header 'Content-Type: application/ld+json' \
--header 'Link: <https://fiware.github.io/data-models/full-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
--data-raw '{
	"type": "Subscription",
	"entities": [{
		"type": "House"
	}],
	"notification": {
		"endpoint": {
			"uri": "http://consumer:8080/notify",
			"accept": "application/json"
		}
	}
}'
```

2. Create a `House` entity. Notice `House` entity has a `hasRoom` relationship with two different objects of `Room` entity (one-to-may relationship or multi-relationship).
```
curl --location --request POST 'http://localhost:9090/ngsi-ld/v1/entities/' \
--header 'Content-Type: application/json' \
--header 'Link: <https://fiware.github.io/data-models/full-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
--data-raw '{
        "id": "smartcity:houses:house1",
        "type": "House",
        "hasRoom": [{
                "type": "Relationship",
                "object": "urn:ngsi-ld:house1:Room:room1",
                "datasetId": "urn:ngsi-ld:Relationship:1"
        },
        {
                "type": "Relationship",
                "object": "urn:ngsi-ld:house1:Room:room2",
                "datasetId": "urn:ngsi-ld:Relationship:2"
        }]
}'
```

3. A Notification is successfully sent to the specified endpoint, with the `hasRoom` multi-relationship pointing at the two `Room` entity objects (i.e., the `urn:ngsi-ld:house1:Room:room1` and `urn:ngsi-ld:house1:Room:room2` objects with their corresponding datasetId appear in the notification).
```
{
    'id': 'urn:ngsi-ld:House:house1', 
    'type': 'House', 
    'hasRoom': [{'type': 'Relationship', 'createdAt': '2021-04-04T19:01:43.308000Z', 'datasetId': 'urn:ngsi-ld:Relationship:2', 'object': 'urn:ngsi-ld:house1:Room:room2', 'modifiedAt': '2021-04-04T19:01:43.308000Z'}, 
    {'type': 'Relationship', 'createdAt': '2021-04-04T19:01:43.308000Z', 'datasetId': 'urn:ngsi-ld:Relationship:1', 'object': 'urn:ngsi-ld:house1:Room:room1', 'modifiedAt': '2021-04-04T19:01:43.308000Z'}], 
    'createdAt': '2021-04-04T19:01:43.308000Z', 
    'modifiedAt': '2021-04-04T19:01:43.308000Z', 
    'name': {'type': 'Property', 'createdAt': '2021-04-04T19:01:43.308000Z', 'value': 'Normal House', 'modifiedAt': '2021-04-04T19:01:43.308000Z'}
}
```

4. A partial update operation of a specific relationship on the `hasRoom` multi-relationship (e.g, update the `object` value of the relationship with `urn:ngsi-ld:Relationship:1` `datasetId` to `urn:ngsi-ld:house1:Room:room3`) does trigger a notification.

4.1. - The partial update operation:
```
curl --location --request PATCH 'http://localhost:9090/ngsi-ld/v1/entities/urn:ngsi-ld:House:house1'/attrs/hasRoom \
--header 'Content-Type: application/json' \
--header 'Link: <https://fiware.github.io/data-models/full-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
--data-raw '{
    "type": "Relationship",
    "datasetId": "urn:ngsi-ld:Relationship:1",
    "object": "urn:ngsi-ld:house1:Room:room3"
}'
```
4.2. -  The notification triggered: 
```
{
    'id': 'urn:ngsi-ld:House:house1', 
    'type': 'House', 
    'hasRoom': [{'type': 'Relationship', 'createdAt': '2021-04-04T19:22:53.718000Z', 'datasetId': 'urn:ngsi-ld:Relationship:2', 'object': 'urn:ngsi-ld:house1:Room:room2', 'modifiedAt': '2021-04-04T19:22:53.718000Z'}, 
    {'type': 'Relationship', 'createdAt': '2021-04-04T19:22:53.718000Z', 'datasetId': 'urn:ngsi-ld:Relationship:1', 'object': 'urn:ngsi-ld:house1:Room:room3', 'modifiedAt': '2021-04-04T19:22:53.718000Z'}], 
    'createdAt': '2021-04-04T19:22:53.718000Z', 
    'modifiedAt': '2021-04-04T19:23:12.626000Z', 
    'name': {'type': 'Property', 'createdAt': '2021-04-04T19:22:53.718000Z', 'value': 'Normal House', 'modifiedAt': '2021-04-04T19:22:53.718000Z'}
}
```

5. However, if you issue an Append operation on the `hasRoom` multi-relationship (e.g, append three relationships with `Room` entity objects), this does not trigger a notification. An example of the Append operation:
```
curl --location --request PATCH 'http://localhost:9090/ngsi-ld/v1/entities/urn:ngsi-ld:House:house1'/attrs \
--header 'Content-Type: application/json' \
--header 'Link: <https://fiware.github.io/data-models/full-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
--data-raw '{
    "hasRoom": [
        {
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:1",
            "object": "urn:ngsi-ld:house1:Room:room1"
        },
        {
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:2",
            "object": "urn:ngsi-ld:house1:Room:room2"
        },
        {
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:3",
            "object": "urn:ngsi-ld:house1:Room:room3"
        }
    ]
}'
```

6. Even though the notification is not triggered, If you query the `House` entity by its `id` , the full array of objects appears in the `hasRoom` multi-relationship. 
```
curl --location --request GET 'http://localhost:9090/ngsi-ld/v1/entities/urn:ngsi-ld:House:house1' \
--header 'Accept: application/ld+json' \
--header 'Link: <https://fiware.github.io/data-models/full-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
--data-raw '{
    "id": "urn:ngsi-ld:House:house1",
    "type": "House",
    "hasRoom": [
        {
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:1",
            "object": "urn:ngsi-ld:house1:Room:room1"
        },
        {
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:2",
            "object": "urn:ngsi-ld:house1:Room:room2"
        },
        {
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:3",
            "object": "urn:ngsi-ld:house1:Room:room3"
        }
    ],
    "name": {
        "type": "Property",
        "value": "Normal House"
    }
}'
```

7. Besides this, I have tried to make use of the Append operation together with the `noOverwrite` option parameter  to add a new relationship with another `Room` entity object on the `hasRoom` multi-relationship (e.g., starting from the entity House created in the step two, append a new relationship with `urn:ngsi-ld:house1:Room:room3` `object` value and `urn:ngsi-ld:Relationship:3` `datasetId` value). This operation does trigger a notification, but overwrites the existing `hasRoom` multi-relationship.

7.1 -  The append operation with the `noOverwrite` option parameter: 
```
curl --location --request PATCH 'http://localhost:9090/ngsi-ld/v1/entities/urn:ngsi-ld:House:house1'/attrs?options=noOverwrite \
--header 'Content-Type: application/json' \
--header 'Link: <https://fiware.github.io/data-models/full-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
--data-raw '{
    "hasRoom": [{
            "type": "Relationship",
            "datasetId": "urn:ngsi-ld:Relationship:3",
            "object": "urn:ngsi-ld:house1:Room:room3"
    }]
}'
```
7.2 - The notification triggered:
```
{
    'id': 'urn:ngsi-ld:House:house1', 
    'type': 'House', 
    'hasRoom': {'type': 'Relationship', 'createdAt': '2021-04-04T20:26:25.340000Z', 'datasetId': 'urn:ngsi-ld:Relationship:3', 'object': 'urn:ngsi-ld:house1:Room:room3', 'modifiedAt': '2021-04-04T20:26:43.401000Z'}, 
    'createdAt': '2021-04-04T20:26:25.340000Z', 
    'modifiedAt': '2021-04-04T20:26:43.401000Z', 
    'name': {'type': 'Property', 'createdAt': '2021-04-04T20:26:25.340000Z', 'value': 'Normal House', 'modifiedAt': '2021-04-04T20:26:25.340000Z'}
}
````

Correct me if I'm not right, but I understand that the `noOverwrite` option parameter should allow the entire `hasRoom` multi-relationship to not be overwritten, and instead add a new relationship to the existing ones (i.e., append the new relationship with `urn:ngsi-ld:house1:Room:room3` entity `object` to the existing ones relationships with `urn:ngsi-ld:house1:Room:room1` and `urn:ngsi-ld:house1:Room:room2` entity objects on the `hasRoom` multi-relationship).

Many thanks in advance!

Daniel

**Comment by amitrnec** [comment]
Hi,
I think i got the problem pined down. PR should be incoming soon.

Thanks,
Amit

**Comment by amitrnec** [comment]
Fixed this issue in #222
